File: tecnologia-compose/app/src/main/java/com/gmail/juliocesar64914/interfazconcompose/pantallas/Seccion4Listas.kt
package com.gmail.juliocesar64914.interfazconcompose.pantallas

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.GridItemSpan
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ViewList
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.filled.DeleteSweep
import androidx.compose.material.icons.filled.GridView
import androidx.compose.material.icons.filled.Inventory2
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material.icons.filled.ViewAgenda
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.PrimaryTabRow
import androidx.compose.material3.Surface
import androidx.compose.material3.SwipeToDismissBox
import androidx.compose.material3.SwipeToDismissBoxValue
import androidx.compose.material3.Tab
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.material3.rememberSwipeToDismissBoxState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.gmail.juliocesar64914.interfazconcompose.datos.CATEGORIA_USUARIO
import com.gmail.juliocesar64914.interfazconcompose.datos.CatalogoViewModel
import com.gmail.juliocesar64914.interfazconcompose.datos.ElementoLista
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

private data class Pestana(val titulo: String, val icono: ImageVector)

private val PESTANAS = listOf(
    Pestana("Lista", Icons.AutoMirrored.Filled.ViewList),
    Pestana("Cuadrícula", Icons.Filled.GridView),
    Pestana("Encabezados", Icons.Filled.ViewAgenda)
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun Seccion4Listas(viewModel: CatalogoViewModel) {
    val pagerState = rememberPagerState(pageCount = { PESTANAS.size })
    val scope = rememberCoroutineScope()
    var detalle by remember { mutableStateOf<ElementoLista?>(null) }

    Column(Modifier.fillMaxSize()) {
        // Documentación de las pestañas
        Guia(
            nombre = "Pestañas con contenido deslizable",
            texto = "Organizan vistas relacionadas en páginas. Toca una pestaña o desliza " +
                    "a la izquierda o a la derecha para cambiar de página.",
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 8.dp)
        )
        PrimaryTabRow(selectedTabIndex = pagerState.currentPage) {
            PESTANAS.forEachIndexed { indice, pestana ->
                Tab(
                    selected = pagerState.currentPage == indice,
                    onClick = { scope.launch { pagerState.animateScrollToPage(indice) } },
                    text = { Text(pestana.titulo) },
                    icon = { Icon(pestana.icono, contentDescription = null) }
                )
            }
        }
        HorizontalPager(
            state = pagerState,
            modifier = Modifier.fillMaxSize()
        ) { pagina ->
            when (pagina) {
                0 -> PaginaLista(viewModel, onVerDetalle = { detalle = it })
                1 -> PaginaCuadricula(viewModel, onVerDetalle = { detalle = it })
                else -> PaginaEncabezados(viewModel, onVerDetalle = { detalle = it })
            }
        }
    }

    // Detalle del elemento seleccionado
    detalle?.let { elemento ->
        AlertDialog(
            onDismissRequest = { detalle = null },
            icon = { Avatar(elemento, 56.dp) },
            title = { Text(elemento.titulo) },
            text = {
                Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                    Text("Categoría: ${elemento.categoria}")
                    Text("Descripción: ${elemento.subtitulo}")
                    Text("Identificador: #${elemento.id}")
                }
            },
            confirmButton = {
                TextButton(onClick = { detalle = null }) { Text("Cerrar") }
            },
            dismissButton = {
                TextButton(onClick = {
                    viewModel.eliminarElemento(elemento)
                    detalle = null
                }) { Text("Eliminar") }
            }
        )
    }
}

// ================= Pestaña 1: lista vertical =================
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun PaginaLista(viewModel: CatalogoViewModel, onVerDetalle: (ElementoLista) -> Unit) {
    var actualizando by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    PullToRefreshBox(
        isRefreshing = actualizando,
        onRefresh = {
            scope.launch {
                actualizando = true
                delay(1500) // simula una descarga
                viewModel.reiniciarLista()
                actualizando = false
            }
        },
        modifier = Modifier.fillMaxSize()
    ) {
        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            contentPadding = PaddingValues(bottom = 16.dp)
        ) {
            item {
                ElevatedCard(Modifier.fillMaxWidth().padding(16.dp)) {
                    Column(
                        Modifier.padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(10.dp)
                    ) {
                        Guia(
                            "Lista vertical",
                            "Muestra elementos uno debajo de otro y solo dibuja los visibles, " +
                                    "por lo que funciona bien con listas largas."
                        )
                        Guia(
                            "Ver detalle",
                            "Toca un elemento para abrir su información completa."
                        )
                        Guia(
                            "Deslizar para eliminar",
                            "Arrastra un elemento hacia un lado para borrarlo de la lista."
                        )
                        Guia(
                            "Actualizar arrastrando hacia abajo",
                            "Desde el inicio de la lista, jala hacia abajo para recargar los datos."
                        )
                        Guia(
                            "Estado vacío",
                            "Cuando no hay elementos se muestra un mensaje con ilustración. " +
                                    "Pruébalo con el botón de vaciar."
                        )
                        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            OutlinedButton(
                                onClick = { viewModel.vaciarLista() },
                                enabled = viewModel.elementos.isNotEmpty()
                            ) {
                                Icon(Icons.Filled.DeleteSweep, contentDescription = null)
                                Spacer(Modifier.size(8.dp))
                                Text("Vaciar lista")
                            }
                        }
                        Text(
                            "${viewModel.elementos.size} elementos",
                            style = MaterialTheme.typography.labelLarge,
                            color = MaterialTheme.colorScheme.primary
                        )
                    }
                }
            }
            if (viewModel.elementos.isEmpty()) {
                item { EstadoVacio(onRestaurar = { viewModel.reiniciarLista() }) }
            } else {
                items(viewModel.elementos, key = { it.id }) { elemento ->
                    FilaDeslizable(
                        elemento = elemento,
                        onClick = { onVerDetalle(elemento) },
                        onEliminar = { viewModel.eliminarElemento(elemento) }
                    )
                    HorizontalDivider()
                }
            }
        }
    }
}

@Composable
private fun FilaDeslizable(
    elemento: ElementoLista,
    onClick: () -> Unit,
    onEliminar: () -> Unit
) {
    val estado = rememberSwipeToDismissBoxState()

    // Cuando el elemento termina de deslizarse, se elimina
    LaunchedEffect(estado.currentValue) {
        if (estado.currentValue != SwipeToDismissBoxValue.Settled) onEliminar()
    }

    SwipeToDismissBox(
        state = estado,
        backgroundContent = {
            val alineacion =
                if (estado.dismissDirection == SwipeToDismissBoxValue.StartToEnd) Alignment.CenterStart
                else Alignment.CenterEnd
            Box(
                Modifier
                    .fillMaxSize()
                    .background(MaterialTheme.colorScheme.errorContainer)
                    .padding(horizontal = 24.dp),
                contentAlignment = alineacion
            ) {
                Icon(
                    Icons.Filled.Delete,
                    contentDescription = "Eliminar",
                    tint = MaterialTheme.colorScheme.onErrorContainer
                )
            }
        }
    ) {
        ListItem(
            headlineContent = { Text(elemento.titulo) },
            supportingContent = { Text(elemento.subtitulo) },
            leadingContent = { Avatar(elemento, 40.dp) },
            modifier = Modifier.clickable(onClick = onClick)
        )
    }
}

// ================= Pestaña 2: cuadrícula =================
@Composable
private fun PaginaCuadricula(viewModel: CatalogoViewModel, onVerDetalle: (ElementoLista) -> Unit) {
    LazyVerticalGrid(
        columns = GridCells.Adaptive(minSize = 104.dp),
        modifier = Modifier.fillMaxSize(),
        contentPadding = PaddingValues(16.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        item(span = { GridItemSpan(maxLineSpan) }) {
            Guia(
                "Cuadrícula de elementos",
                "Acomoda los elementos en filas y columnas. El número de columnas se " +
                        "ajusta al ancho de la pantalla. Toca una tarjeta para ver su detalle."
            )
        }
        if (viewModel.elementos.isEmpty()) {
            item(span = { GridItemSpan(maxLineSpan) }) {
                EstadoVacio(onRestaurar = { viewModel.reiniciarLista() })
            }
        } else {
            items(viewModel.elementos, key = { it.id }) { elemento ->
                Card(onClick = { onVerDetalle(elemento) }) {
                    Column(
                        Modifier
                            .fillMaxWidth()
                            .padding(12.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(6.dp)
                    ) {
                        Avatar(elemento, 48.dp)
                        Text(
                            elemento.titulo,
                            style = MaterialTheme.typography.titleSmall,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                        Text(
                            elemento.categoria,
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
            }
        }
    }
}

// ================= Pestaña 3: lista con encabezados =================
@OptIn(ExperimentalFoundationApi::class)
@Composable
private fun PaginaEncabezados(viewModel: CatalogoViewModel, onVerDetalle: (ElementoLista) -> Unit) {
    val grupos = viewModel.elementos.groupBy { it.categoria }

    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        contentPadding = PaddingValues(bottom = 16.dp)
    ) {
        item {
            Guia(
                "Lista con encabezados de sección",
                "Combina dos tipos de elemento: encabezados que agrupan por categoría y " +
                        "filas de contenido. El encabezado se queda fijo arriba mientras recorres su grupo.",
                modifier = Modifier.padding(16.dp)
            )
        }
        if (grupos.isEmpty()) {
            item { EstadoVacio(onRestaurar = { viewModel.reiniciarLista() }) }
        }
        grupos.forEach { (categoria, lista) ->
            stickyHeader(key = "encabezado-$categoria", contentType = "encabezado") {
                Surface(
                    color = MaterialTheme.colorScheme.secondaryContainer,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Text(
                        "$categoria (${lista.size})",
                        style = MaterialTheme.typography.titleSmall,
                        fontWeight = FontWeight.Bold,
                        color = MaterialTheme.colorScheme.onSecondaryContainer,
                        modifier = Modifier.padding(horizontal = 16.dp, vertical = 10.dp)
                    )
                }
            }
            items(lista, key = { it.id }, contentType = { "fila" }) { elemento ->
                ListItem(
                    headlineContent = { Text(elemento.titulo) },
                    supportingContent = { Text(elemento.subtitulo) },
                    leadingContent = { Avatar(elemento, 40.dp) },
                    modifier = Modifier.clickable { onVerDetalle(elemento) }
                )
            }
        }
    }
}

// ================= Componentes compartidos =================

/** Nombre y explicación breve de un elemento, en formato compacto. */
@Composable
private fun Guia(nombre: String, texto: String, modifier: Modifier = Modifier) {
    Column(modifier) {
        Text(
            nombre,
            style = MaterialTheme.typography.titleSmall,
            color = MaterialTheme.colorScheme.primary
        )
        Text(
            texto,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

/** Círculo con la inicial del elemento; el color depende de la categoría. */
@Composable
private fun Avatar(elemento: ElementoLista, tamano: Dp) {
    val (fondo: Color, texto: Color) = when (elemento.categoria) {
        "Frutas" -> MaterialTheme.colorScheme.primaryContainer to MaterialTheme.colorScheme.onPrimaryContainer
        "Verduras" -> MaterialTheme.colorScheme.tertiaryContainer to MaterialTheme.colorScheme.onTertiaryContainer
        CATEGORIA_USUARIO -> MaterialTheme.colorScheme.secondaryContainer to MaterialTheme.colorScheme.onSecondaryContainer
        else -> MaterialTheme.colorScheme.surfaceVariant to MaterialTheme.colorScheme.onSurfaceVariant
    }
    Surface(shape = CircleShape, color = fondo, modifier = Modifier.size(tamano)) {
        Box(contentAlignment = Alignment.Center) {
            Text(
                elemento.titulo.take(1).uppercase(),
                color = texto,
                style = if (tamano >= 48.dp) MaterialTheme.typography.titleLarge
                else MaterialTheme.typography.titleMedium
            )
        }
    }
}

/** Estado vacío con ilustración, mensaje y acción para recuperar los datos. */
@Composable
private fun EstadoVacio(onRestaurar: () -> Unit) {
    Column(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 32.dp, vertical = 40.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Surface(
            shape = CircleShape,
            color = MaterialTheme.colorScheme.surfaceVariant,
            modifier = Modifier.size(120.dp)
        ) {
            Box(contentAlignment = Alignment.Center) {
                Icon(
                    Icons.Filled.Inventory2,
                    contentDescription = null,
                    modifier = Modifier.size(56.dp),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
        Text("La lista está vacía", style = MaterialTheme.typography.titleMedium)
        Text(
            "No hay elementos para mostrar. Restaura la lista original o agrega uno " +
                    "nuevo desde la sección Entrada de texto.",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center
        )
        Spacer(Modifier.height(4.dp))
        Button(onClick = onRestaurar) {
            Icon(Icons.Filled.Refresh, contentDescription = null)
            Spacer(Modifier.size(8.dp))
            Text("Restaurar lista")
        }
    }
}

File: tecnologia-flutter/lib/pantallas/seccion4_listas.dart
import 'package:flutter/material.dart';

import '../datos/catalogo_estado.dart';

/// Sección 4: pestañas (TabBar) + páginas deslizables (TabBarView).
class Seccion4Listas extends StatelessWidget {
  const Seccion4Listas({super.key});

  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 3,
      child: Column(
        children: [
          const Padding(
            padding: EdgeInsets.fromLTRB(16, 8, 16, 8),
            child: _Guia(
              nombre: 'Pestañas con contenido deslizable',
              texto: 'Organizan vistas relacionadas en páginas. Toca una pestaña o desliza a '
                  'la izquierda o a la derecha para cambiar de página.',
            ),
          ),
          // Pestañas secundarias: se distinguen de las pestañas principales de la app
          const TabBar.secondary(
            tabs: [Tab(text: 'Lista'), Tab(text: 'Cuadrícula'), Tab(text: 'Encabezados')],
          ),
          const Expanded(
            child: TabBarView(
              children: [_PaginaLista(), _PaginaCuadricula(), _PaginaEncabezados()],
            ),
          ),
        ],
      ),
    );
  }
}

// ================= Pestaña 1: lista vertical =================
class _PaginaLista extends StatelessWidget {
  const _PaginaLista();

  @override
  Widget build(BuildContext context) {
    final estado = EstadoScope.of(context);
    final elementos = estado.elementos;
    final tema = Theme.of(context);

    // RefreshIndicator: jalar hacia abajo para actualizar
    return RefreshIndicator(
      onRefresh: () async {
        await Future.delayed(const Duration(milliseconds: 1500)); // simula una descarga
        estado.reiniciarLista();
      },
      child: ListView(
        // Permite jalar aunque la lista esté vacía
        physics: const AlwaysScrollableScrollPhysics(),
        padding: const EdgeInsets.only(bottom: 16),
        children: [
          Card(
            margin: const EdgeInsets.all(16),
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                spacing: 12,
                children: [
                  const _Guia(
                    nombre: 'Lista vertical',
                    texto: 'Muestra elementos uno debajo de otro y solo dibuja los visibles, '
                        'por lo que funciona bien con listas largas.',
                  ),
                  const _Guia(nombre: 'Ver detalle', texto: 'Toca un elemento para abrir su información completa.'),
                  const _Guia(
                    nombre: 'Deslizar para eliminar',
                    texto: 'Arrastra un elemento hacia un lado para borrarlo de la lista.',
                  ),
                  const _Guia(
                    nombre: 'Actualizar arrastrando hacia abajo',
                    texto: 'Desde el inicio de la lista, jala hacia abajo para recargar los datos.',
                  ),
                  const _Guia(
                    nombre: 'Estado vacío',
                    texto: 'Cuando no hay elementos se muestra un mensaje con ilustración. '
                        'Pruébalo con el botón de vaciar.',
                  ),
                  OutlinedButton.icon(
                    onPressed: elementos.isEmpty ? null : estado.vaciarLista,
                    icon: const Icon(Icons.delete_sweep),
                    label: const Text('Vaciar lista'),
                  ),
                  Text(
                    '${elementos.length} elementos',
                    style: tema.textTheme.labelLarge?.copyWith(color: tema.colorScheme.primary),
                  ),
                ],
              ),
            ),
          ),
          if (elementos.isEmpty)
            const _EstadoVacio()
          else
            for (final elemento in elementos) ...[
              // Dismissible: deslizar para eliminar
              Dismissible(
                key: ValueKey(elemento.id),
                background: const _FondoEliminar(alineacion: Alignment.centerLeft),
                secondaryBackground: const _FondoEliminar(alineacion: Alignment.centerRight),
                onDismissed: (_) => estado.eliminarElemento(elemento),
                child: ListTile(
                  leading: _Avatar(elemento),
                  title: Text(elemento.titulo),
                  subtitle: Text(elemento.subtitulo),
                  onTap: () => _mostrarDetalle(context, elemento),
                ),
              ),
              const Divider(height: 1),
            ],
        ],
      ),
    );
  }
}

class _FondoEliminar extends StatelessWidget {
  const _FondoEliminar({required this.alineacion});

  final Alignment alineacion;

  @override
  Widget build(BuildContext context) {
    final colores = Theme.of(context).colorScheme;
    return Container(
      color: colores.errorContainer,
      alignment: alineacion,
      padding: const EdgeInsets.symmetric(horizontal: 24),
      child: Icon(Icons.delete, color: colores.onErrorContainer, semanticLabel: 'Eliminar'),
    );
  }
}

// ================= Pestaña 2: cuadrícula =================
class _PaginaCuadricula extends StatelessWidget {
  const _PaginaCuadricula();

  @override
  Widget build(BuildContext context) {
    final elementos = EstadoScope.of(context).elementos;
    final tema = Theme.of(context);

    return CustomScrollView(
      slivers: [
        const SliverPadding(
          padding: EdgeInsets.all(16),
          sliver: SliverToBoxAdapter(
            child: _Guia(
              nombre: 'Cuadrícula de elementos',
              texto: 'Acomoda los elementos en filas y columnas. El número de columnas se '
                  'ajusta al ancho de la pantalla. Toca una tarjeta para ver su detalle.',
            ),
          ),
        ),
        if (elementos.isEmpty)
          const SliverToBoxAdapter(child: _EstadoVacio())
        else
          SliverPadding(
            padding: const EdgeInsets.fromLTRB(16, 0, 16, 16),
            // Cada tarjeta mide como máximo 120 de ancho: las columnas se calculan solas
            sliver: SliverGrid(
              gridDelegate: const SliverGridDelegateWithMaxCrossAxisExtent(
                maxCrossAxisExtent: 120,
                mainAxisSpacing: 12,
                crossAxisSpacing: 12,
                childAspectRatio: 0.9,
              ),
              delegate: SliverChildBuilderDelegate(
                childCount: elementos.length,
                (context, i) {
                  final elemento = elementos[i];
                  return Card.filled(
                    margin: EdgeInsets.zero,
                    clipBehavior: Clip.antiAlias,
                    child: InkWell(
                      onTap: () => _mostrarDetalle(context, elemento),
                      child: Padding(
                        padding: const EdgeInsets.all(8),
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          spacing: 6,
                          children: [
                            _Avatar(elemento, radio: 24),
                            Text(
                              elemento.titulo,
                              maxLines: 1,
                              overflow: TextOverflow.ellipsis,
                              style: tema.textTheme.titleSmall,
                            ),
                            Text(
                              elemento.categoria,
                              style: tema.textTheme.labelSmall
                                  ?.copyWith(color: tema.colorScheme.onSurfaceVariant),
                            ),
                          ],
                        ),
                      ),
                    ),
                  );
                },
              ),
            ),
          ),
      ],
    );
  }
}

// ================= Pestaña 3: lista con encabezados =================
class _PaginaEncabezados extends StatelessWidget {
  const _PaginaEncabezados();

  @override
  Widget build(BuildContext context) {
    final elementos = EstadoScope.of(context).elementos;

    // Agrupa por categoría conservando el orden de aparición
    final grupos = <String, List<ElementoLista>>{};
    for (final e in elementos) {
      grupos.putIfAbsent(e.categoria, () => []).add(e);
    }

    return CustomScrollView(
      slivers: [
        const SliverPadding(
          padding: EdgeInsets.all(16),
          sliver: SliverToBoxAdapter(
            child: _Guia(
              nombre: 'Lista con encabezados de sección',
              texto: 'Combina dos tipos de elemento: encabezados que agrupan por categoría y '
                  'filas de contenido. El encabezado se queda fijo arriba mientras recorres su grupo.',
            ),
          ),
        ),
        if (grupos.isEmpty) const SliverToBoxAdapter(child: _EstadoVacio()),
        for (final grupo in grupos.entries)
          // SliverMainAxisGroup + encabezado fijo (pinned): el encabezado solo
          // permanece arriba mientras se ven los elementos de su grupo
          SliverMainAxisGroup(
            slivers: [
              SliverPersistentHeader(
                pinned: true,
                delegate: _EncabezadoFijo('${grupo.key} (${grupo.value.length})'),
              ),
              SliverList.list(
                children: [
                  for (final elemento in grupo.value)
                    ListTile(
                      leading: _Avatar(elemento),
                      title: Text(elemento.titulo),
                      subtitle: Text(elemento.subtitulo),
                      onTap: () => _mostrarDetalle(context, elemento),
                    ),
                ],
              ),
            ],
          ),
        const SliverToBoxAdapter(child: SizedBox(height: 16)),
      ],
    );
  }
}

class _EncabezadoFijo extends SliverPersistentHeaderDelegate {
  _EncabezadoFijo(this.texto);

  final String texto;

  @override
  double get minExtent => 44;

  @override
  double get maxExtent => 44;

  @override
  Widget build(BuildContext context, double shrinkOffset, bool overlapsContent) {
    final tema = Theme.of(context);
    return Container(
      color: tema.colorScheme.secondaryContainer,
      alignment: Alignment.centerLeft,
      padding: const EdgeInsets.symmetric(horizontal: 16),
      child: Text(
        texto,
        style: tema.textTheme.titleSmall?.copyWith(
          fontWeight: FontWeight.bold,
          color: tema.colorScheme.onSecondaryContainer,
        ),
      ),
    );
  }

  @override
  bool shouldRebuild(_EncabezadoFijo anterior) => anterior.texto != texto;
}

// ================= Componentes compartidos =================

/// Nombre y explicación breve de un elemento, en formato compacto.
class _Guia extends StatelessWidget {
  const _Guia({required this.nombre, required this.texto});

  final String nombre;
  final String texto;

  @override
  Widget build(BuildContext context) {
    final tema = Theme.of(context);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(nombre, style: tema.textTheme.titleSmall?.copyWith(color: tema.colorScheme.primary)),
        Text(
          texto,
          style: tema.textTheme.bodySmall?.copyWith(color: tema.colorScheme.onSurfaceVariant),
        ),
      ],
    );
  }
}

/// Círculo con la inicial; el color depende de la categoría.
class _Avatar extends StatelessWidget {
  const _Avatar(this.elemento, {this.radio = 20});

  final ElementoLista elemento;
  final double radio;

  @override
  Widget build(BuildContext context) {
    final c = Theme.of(context).colorScheme;
    final (fondo, texto) = switch (elemento.categoria) {
      'Frutas' => (c.primaryContainer, c.onPrimaryContainer),
      'Verduras' => (c.tertiaryContainer, c.onTertiaryContainer),
      categoriaUsuario => (c.secondaryContainer, c.onSecondaryContainer),
      _ => (c.surfaceContainerHighest, c.onSurfaceVariant),
    };
    return CircleAvatar(
      radius: radio,
      backgroundColor: fondo,
      foregroundColor: texto,
      child: Text(elemento.titulo.characters.first.toUpperCase()),
    );
  }
}

/// Estado vacío con ilustración, mensaje y acción para recuperar los datos.
class _EstadoVacio extends StatelessWidget {
  const _EstadoVacio();

  @override
  Widget build(BuildContext context) {
    final tema = Theme.of(context);
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 32, vertical: 40),
      child: Column(
        spacing: 12,
        children: [
          CircleAvatar(
            radius: 60,
            backgroundColor: tema.colorScheme.surfaceContainerHighest,
            child: Icon(Icons.inventory_2, size: 56, color: tema.colorScheme.onSurfaceVariant),
          ),
          Text('La lista está vacía', style: tema.textTheme.titleMedium),
          Text(
            'No hay elementos para mostrar. Restaura la lista original o agrega uno nuevo '
            'desde la sección Entrada de texto.',
            textAlign: TextAlign.center,
            style: tema.textTheme.bodyMedium?.copyWith(color: tema.colorScheme.onSurfaceVariant),
          ),
          FilledButton.icon(
            onPressed: EstadoScope.leer(context).reiniciarLista,
            icon: const Icon(Icons.refresh),
            label: const Text('Restaurar lista'),
          ),
        ],
      ),
    );
  }
}

/// Diálogo con el detalle del elemento seleccionado.
void _mostrarDetalle(BuildContext context, ElementoLista elemento) {
  final estado = EstadoScope.leer(context);
  showDialog<void>(
    context: context,
    builder: (contextoDialogo) => AlertDialog(
      icon: _Avatar(elemento, radio: 28),
      title: Text(elemento.titulo),
      content: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        spacing: 4,
        children: [
          Text('Categoría: ${elemento.categoria}'),
          Text('Descripción: ${elemento.subtitulo}'),
          Text('Identificador: #${elemento.id}'),
        ],
      ),
      actions: [
        TextButton(
          onPressed: () {
            estado.eliminarElemento(elemento);
            Navigator.pop(contextoDialogo);
          },
          child: const Text('Eliminar'),
        ),
        TextButton(
          onPressed: () => Navigator.pop(contextoDialogo),
          child: const Text('Cerrar'),
        ),
      ],
    ),
  );
}
